File: src/bumperbot_controller/bumperbot_controller/simple_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import Twist
import numpy as np

class SimpleController(Node):
    """
    A simple controller class that allows you to set and get parameters.
    """

    def __init__(self):
        super().__init__('simple_controller')

        self.declare_parameter('wheel_radius', 0.033)
        self.declare_parameter('wheel_separation', 0.17)


        self.wheel_radius = self.get_parameter('wheel_radius').get_parameter_value().double_value
        self.wheel_separation = self.get_parameter('wheel_separation').get_parameter_value().double_value


        self.get_logger().info(f'Wheel radius: {self.wheel_radius} \n'
                            f'Wheel separation: {self.wheel_separation}')
        

        self.wheel_cmd_pub = self.create_publisher(Float64MultiArray, 'simple_velocity_controller/commands', 10)
        self.vel_subscriber = self.create_subscription(Twist, 'bumperbot_controller/cmd_vel', self.vel_callback, 10)

        # Conversion matrix for wheel speeds
        self.speed_conversion_ = np.array([[self.wheel_radius/2, self.wheel_radius/2], 
                                           [self.wheel_radius/self.wheel_separation, -self.wheel_radius/self.wheel_separation]
                                           ])  


    
    def vel_callback(self, msg):

        robot_speed = np.array([msg.twist.linear.x],
                               [msg.twist.angular.z])

        # Calculate the wheel velocities
        wheel_speed = np.matmul(np.linalg.inv(self.speed_conversion_), robot_speed)

        wheel_speed_cmd = Float64MultiArray()
        wheel_speed_cmd.data = [wheel_speed[1, 0], wheel_speed[0, 0]]

        # Publish the wheel commands
        self.wheel_cmd_pub.publish(wheel_speed_cmd)


def main(args=None):
    rclpy.init(args = args)
    simple_controller = SimpleController()
    rclpy.spin(simple_controller)
    simple_controller.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_controller/src/simple_controller.cpp
#include "bumperbot_controller/simple_controller.hpp"
#include <Eigen/Geometry>


SimpleController :: SimpleController(const std::string &name) :Node(name)

{
    declare_parameter("wheel_radius", 0.033);
    declare_parameter("wheel_separation", 0.17);


    wheel_radius = get_parameter("wheel_radius").as_double();
    wheel_separation = get_parameter("wheel_separation").as_double();

    RCLCPP_INFO(get_logger(), "Wheel radius: %f", wheel_radius);
    RCLCPP_INFO(get_logger(), "Wheel separation: %f", wheel_separation);

    wheel_cmd_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>("/simple_velocity_controller/commands", 10);
    vel_sub_ = create_subscription<geometry_msgs::msg::TwistStamped>(
        "/bumperbot_controller/cmd_vel", 10, std::bind(&SimpleController::velCallback, this, std::placeholders::_1));

    speed_conversion_  << wheel_radius / 2.0, wheel_radius / 2.0, 
                         wheel_radius / wheel_separation, -wheel_radius / wheel_separation;
}


void SimpleController :: velCallback(const geometry_msgs::msg::TwistStamped & msg)
{
    Eigen ::Vector2d robot_speed(msg.twist.linear.x, msg.twist.angular.z);

    Eigen ::Vector2d wheel_speed = speed_conversion_.inverse() * robot_speed;

    std_msgs::msg::Float64MultiArray wheel_speed_msg;

    wheel_speed_msg.data.push_back(wheel_speed.coeff(1));
    wheel_speed_msg.data.push_back(wheel_speed.coeff(0));

    wheel_cmd_pub_->publish(wheel_speed_msg);
}

int main (int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleController>("simple_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
} 